File: leetcode/lc-us/cpp/1813.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    bool areSentencesSimilar(string sentence1, string sentence2) {
        if (sentence1.length() < sentence2.length()) return areSentencesSimilar(sentence2, sentence1);
        deque<string> dq1, dq2;
        stringstream ss1(sentence1), ss2(sentence2);
        string word;
        while (ss1 >> word) dq1.push_back(word);
        while (ss2 >> word) dq2.push_back(word);
        while (true) {
            while (!dq1.empty() && !dq2.empty() && dq1.front() == dq2.front()) {
                dq1.pop_front();
                dq2.pop_front();
            }
            while (!dq1.empty() && !dq2.empty() && dq1.back() == dq2.back()) {
                dq1.pop_back();
                dq2.pop_back();
            }
            break;
        }
        return dq2.empty();
    }
};

File: leetcode/lc-us/py/1813.py
class Solution:
    def areSentencesSimilar(self, sentence1: str, sentence2: str) -> bool:
        if sentence1 == sentence2:
            return True
        if len(sentence1) < len(sentence2):
            sentence1, sentence2 = sentence2, sentence1
        words1, words2 = sentence1.split(), sentence2.split()
        if len(words1) < len(words2):
            return False
        i = 0
        while i < len(words2) and words1[i] == words2[i]:
            i += 1
        j, k = len(words1) - 1, len(words2) - 1
        while k >= i:
            if words1[j] != words2[k]:
                return False
            j -= 1
            k -= 1
        return True
